userguide/tutorials/invoice_examples.adoc

```diff
@@ -0,0 +1,268 @@
+= Invoice Examples
+
+Whenever there is a change in the system (a recurring subscription billing, an external charge, a change in plan, and so on), Kill Bill generates an invoice. An invoice consists of one or more invoice items of different types. In addition, an invoice has the `balance` and `chargedAmount` attributes which are computed each time an invoice is generated. This document aims to explain these computations in different invoicing scenarios.
+
+== Prerequisites
+
+* Ensure that you have gone through the https://docs.killbill.io/latest/userguide_subscription.html#components-invoicing[_Subscription User Guide_] and are familiar with the basic invoicing concepts. 
+
+* Ensure that you are familiar with the invoice related terminology explained in the https://docs.killbill.io/latest/Kill-Bill-Glossary.html[_Kill Bill Glossary_].  
+
+== More about Charged Amount and Balance
+
+Before we look at each scenario, let us understand how `charged amount` and `balance` are calculated. 
+
+=== Charged Amount/Invoice Amount/Amount
+
+The `charged amount` (also known as `invoice amount`, `invoice total` or simply `amount`) specifies the amount that the customer is charged as part of an invoice. It is computed by adding up the amounts on invoice items of type `FIXED`, `RECURRING`, `EXTERNAL_CHARGE`, `USAGE`, `TAX`, `ITEM_ADJ`, `REPAIR_ADJ`, `PARENT_SUMMARY` and `CREDIT_ADJ`. Note that `PARENT_SUMMARY` invoice items are only applicable in the case of https://docs.killbill.io/latest/ha.html[hierarchical accounts] and will be skipped in the rest of this document.
+
+So, `charged amount` can be expressed as follows:
+
+[source,java]
+CHARGED AMOUNT = FIXED + RECURRING + EXTERNAL_CHARGE + USAGE + TAX + ITEM_ADJ + REPAIR_ADJ + CREDIT_ADJ
+
+[NOTE]
+*Note:* In the case of an invoice credit, the `CREDIT_ADJ` invoice items are not taken into account while calculating the charged amount. This is to prevent getting a negative amount as explained in the invoice credit scenario below.
+
+=== Invoice Balance
+
+`Invoice balance` specifies the amount that the customer owes as part of an invoice. It is positive if the customer owes something or zero otherwise. It is calculated by subtracting the sum of the `paid amount` (amount paid by the customer) and `refunded amount` (amount refunded to the customer) from the sum of the `charged amount` and internal credit adjustment (`CBA_ADJ` items) amount. 
+
+So, `Invoice Balance` can be expressed as follows:
+
+[source,java]
+INVOICE BALANCE = CHARGED AMOUNT + SUM(CBA_ADJ) - (AMOUNT PAID + AMOUNT REFUNDED) 
```

Comment:
That is still different than what I had:
`INVOICE BALANCE = CHARGED AMOUNT - (AMOUNT PAID - AMOUNT REFUNDED) + SUM(CBA_ADJ)`

I suppose it really depends if `AMOUNT REFUNDED` is positive or negative. In 'my' formula this is expected to be positive but in yours, this is expected to be negative. Could you double check what the code is doing?

Also I think we should add for all these fields what is the expectation. Example: 
* `CREDIT_ADJ` always negative
* `FIXED`, `RECURRING`, ... always positive
* ...
*

Reply:
I checked the code and the formula that I've written looks correct to me. The [InvoiceCalculatorUtils.computeInvoiceAmountRefunded](https://github.com/killbill/killbill/blob/d6a0c501af82b2affa706cf3e22f573cd0fc28da/invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java#L228) method returns a **negative** value corresponding to the `refund` (When a refund is performed, there is a record in the `invoice_payments` table of type `REFUND` and a negative value). The [InvoiceCalculatorUtils.computeRawInvoiceBalance](https://github.com/killbill/killbill/blob/d6a0c501af82b2affa706cf3e22f573cd0fc28da/invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java#L100) method then adds the `amount refunded` (negative value) to `amount paid`. 

With regards to adding whether each field is positive or negative, this information is already present in the [subscription billing document](https://docs.killbill.io/latest/userguide_subscription.html#components-invoicing-overview). So, for now, I've just added a note with a link to this document to avoid duplication. I feel the subscription billing document anyway needs some reorganization, so when we do that we can give more thought if we want to keep this information there or move it elsewhere. But if you think this information is required in this document, I can add it.